File: src/main/java/engine/glv2/pipeline/LensFlares.java
/*
 *
 * Copyright (C) 2015-2020 Anarchy Engine Open Source Contributors (see CONTRIBUTORS.md)
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 */

package engine.glv2.pipeline;

import static org.lwjgl.opengl.GL13C.GL_TEXTURE0;
import static org.lwjgl.opengl.GL13C.GL_TEXTURE1;

import engine.glv2.objects.Texture;
import engine.glv2.pipeline.shaders.LensFlaresShader;
import engine.glv2.v2.DeferredPass;
import engine.glv2.v2.DeferredPipeline;
import engine.glv2.v2.RendererData;
import engine.resources.ResourcesManager;

public class LensFlares extends DeferredPass<LensFlaresShader> {

	private Texture lensColor;

	public LensFlares() {
		super("LensFlares");
	}

	@Override
	public void init(int width, int height) {
		super.init(width, height);
		lensColor = ResourcesManager.loadTextureMisc("assets/textures/lens/lens_color.png", null).get();
	}

	@Override
	protected LensFlaresShader setupShader() {
		return new LensFlaresShader();
	}

	@Override
	protected void setupTextures(RendererData rnd, DeferredPipeline dp, Texture[] auxTex) {
		auxTex[1].active(GL_TEXTURE0);
		lensColor.active(GL_TEXTURE1);
		auxTex[1] = auxTex[0];
	}

	@Override
	public void dispose() {
		super.dispose();
		lensColor.dispose();
	}

}

File: src/main/resources/assets/shaders/deferred/LensFlares.fs
/*

Copyright (C) 2015-2020 Anarchy Engine Open Source Contributors (see CONTRIBUTORS.md)

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.

 */

in vec2 textureCoords;

out vec4 out_Color;

uniform sampler2D bloom;
uniform sampler2D lensColor;

uniform int useLensFlares;

const int ghosts = 5;
const float ghostDispersal = 0.4;
const float haloWidth = 0.47;
const float distortion = 1.5;

vec3 textureDistorted(sampler2D tex, vec2 texcoord, vec2 direction, vec3 distortion) {
	return vec3(texture(tex, texcoord + direction * distortion.r).r,
				texture(tex, texcoord + direction * distortion.g).g,
				texture(tex, texcoord + direction * distortion.b).b);
}

void main() {
	vec3 result = vec3(0.0);
	if (useLensFlares == 1) {
		vec2 texcoord = -textureCoords + vec2(1.0);
		vec2 ghostVec = (vec2(0.5) - texcoord) * ghostDispersal;
		vec2 texelSize = 1.0 / vec2(textureSize(bloom, 0));
		vec3 distortion = vec3(-texelSize.x * distortion, 0.0, texelSize.x * distortion);
		vec2 direction = normalize(ghostVec);
		for (int i = 0; i < ghosts; ++i) {
			vec2 offset = texcoord + ghostVec * float(i);
			float weight = length(vec2(0.5) - offset) / length(vec2(0.5));
			weight = pow(1.0 - weight, 10.0);
			result += textureDistorted(bloom, offset, direction, distortion) * weight;
		}
		vec2 haloVec = normalize(ghostVec) * haloWidth;
		float weight = length(vec2(0.5) - fract(texcoord + haloVec)) / length(vec2(0.5));
		weight = pow(1.0 - weight, 5.0);
		result += textureDistorted(bloom, texcoord + haloVec, direction, distortion) * weight;
		result *= texture(lensColor, vec2(length(vec2(0.5) - texcoord) / length(vec2(0.5)), 0.0)).rgb;
	}
	out_Color.rgb = result;
	out_Color.a = 0.0;
}